lnwire/channel_update_2.go

```diff
@@ -0,0 +1,510 @@
+	)
+
+	records := []tlv.Record{
+		tlv.MakePrimitiveRecord(ChanUpdate2ChainHashType, &chainHash),
+		scidRecordProducer.Record(),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2BlockHeightType, &c.BlockHeight,
+		),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2DisableFlagsType, &disableFlags,
+		),
+		directionRecordProducer.Record(),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2CLTVExpiryDeltaType, &c.CLTVExpiryDelta,
+		),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2HTLCMinMsatType, &htlcMin,
+		),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2HTLCMaxMsatType, &htlcMax,
+		),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2FeeBaseMsatType, &c.FeeBaseMsat,
+		),
+		tlv.MakePrimitiveRecord(
+			ChanUpdate2FeeProportionalMillionthsType,
+			&c.FeeProportionalMillionths,
+		),
+	}
+
+	typeMap, err := tlvRecords.ExtractRecords(records...)
+	if err != nil {
+		return err
+	}
+
+	// By default, the chain-hash is the bitcoin mainnet genesis block hash.
+	c.ChainHash = *chaincfg.MainNetParams.GenesisHash
+	if _, ok := typeMap[ChanUpdate2ChainHashType]; ok {
+		c.ChainHash = chainHash
+	}
+
+	if _, ok := typeMap[ChanUpdate2DisableFlagsType]; ok {
+		c.DisabledFlags = ChanUpdateDisableFlags(disableFlags)
+	}
+
+	if _, ok := typeMap[ChanUpdate2SCIDType]; ok {
+		c.ShortChannelID = scidRecordProducer.ShortChannelID
+	}
+
+	if _, ok := typeMap[ChanUpdate2DirectionType]; ok {
+		c.Direction = directionRecordProducer.Bool
+	}
+
+	// If the CLTV expiry delta was not encoded, then set it to the default
+	// value.
+	if _, ok := typeMap[ChanUpdate2CLTVExpiryDeltaType]; !ok {
+		c.CLTVExpiryDelta = defaultCltvExpiryDelta
+	}
+
+	c.HTLCMinimumMsat = defaultHtlcMinMsat
+	if _, ok := typeMap[ChanUpdate2HTLCMinMsatType]; ok {
+		c.HTLCMinimumMsat = MilliSatoshi(htlcMin)
+	}
+
+	if _, ok := typeMap[ChanUpdate2HTLCMaxMsatType]; ok {
+		c.HTLCMaximumMsat = MilliSatoshi(htlcMax)
+	}
+
+	// If the base fee was not encoded, then set it to the default value.
+	if _, ok := typeMap[ChanUpdate2FeeBaseMsatType]; !ok {
+		c.FeeBaseMsat = defaultFeeBaseMsat
+	}
+
+	// If the proportional fee was not encoded, then set it to the default
+	// value.
+	if _, ok := typeMap[ChanUpdate2FeeProportionalMillionthsType]; !ok {
+		c.FeeProportionalMillionths = defaultFeeProportionalMillionths
+	}
+
+	if len(tlvRecords) != 0 {
+		c.ExtraOpaqueData = tlvRecords
+	}
+
+	return nil
+}
+
+// Encode serializes the target ChannelUpdate2 into the passed io.Writer
+// observing the protocol version specified.
+//
+// This is part of the lnwire.Message interface.
+func (c *ChannelUpdate2) Encode(w *bytes.Buffer, _ uint32) error {
+	_, err := w.Write(c.Signature.RawBytes())
+	if err != nil {
+		return err
+	}
+
+	_, err = c.DataToSign()
+	if err != nil {
+		return err
+	}
+
+	return WriteBytes(w, c.ExtraOpaqueData)
+}
+
+// DigestTag returns the tag to be used when signing the digest.
+func (c *ChannelUpdate2) DigestTag() []byte {
+	return MsgTag("channel_announcement_2", "announcement_signature")
```

Comment:
not sure if i'm off base here but the spec draft says it should be `MsgHash("channel_update_2", "bip340_sig", m)`?

Reply:
yeah why would it sign ann sig and chan ann 2?